File: 剑指Offer/4.二维数组中的查找.cpp
class Solution {
public:
    bool findNumberIn2DArray(vector<vector<int>>& matrix, int target) {
        if (matrix.size() == 0 || matrix[0].size() == 0) return false;

        int rows = matrix.size(), cols = matrix[0].size();
        int rightCol = cols - 1;
        int topRow = 0;
        while (topRow < rows && rightCol >= 0) {
            if (matrix[topRow][rightCol] == target) return true;
            if (matrix[topRow][rightCol] > target) {
                rightCol--;
                continue;
            }
            if (matrix[topRow][rightCol] < target) {
                topRow++;
                continue;
            }
        }
        return false;
    }
};

File: 剑指Offer/4.二维数组中的查找.py
class Solution:
    def findNumberIn2DArray(self, matrix: List[List[int]], target: int) -> bool:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        
        rows = len(matrix)
        cols = len(matrix[0])
        rightCol = cols - 1
        topRow = 0

        while topRow < rows and rightCol >= 0:
            if matrix[topRow][rightCol] == target:
                return True
            
            if matrix[topRow][rightCol] > target:
                rightCol -= 1
                continue
            
            if matrix[topRow][rightCol] < target:
                topRow += 1
                continue
        return False